service/history/ndc/workflow_state_replicator.go

```diff
@@ -1150,7 +1170,17 @@ BackfillLoop:
 	for remoteHistoryIterator.HasNext() {
 		historyBlob, err := remoteHistoryIterator.Next()
 		if err != nil {
-			return common.EmptyEventTaskID, err
+			return common.EmptyEventTaskID, nil, err
+		}
+
+		if isStateBased {
+			// If backfill suceeds but later event reapply fails, during task's next retry,
+			// we still need to reapply events that have been stored in local DB.
+			events, err := r.historySerializer.DeserializeEvents(historyBlob.rawHistory)
+			if err != nil {
+				return common.EmptyEventTaskID, nil, err
+			}
+			eventsToReapply = append(eventsToReapply, events...)
```

Comment:
nit: is it possible to add those events directly to ms here instead of buffering entire history in memory, return and then filter?